File: core/src/main/java/com/gumse/gui/Basics/Button.cpp
#include "Button.h"
#include <Essentials/Tools.h>
#include <System/MemoryManagement.h>

#include "../Font/FontManager.h"
#include "../../gum-gui.h"

Button::Button(std::string title) : Button(ivec2(0,0), ivec2(50, 20), title, Gum::GUI::Fonts->getDefaultFont())
{
}

Button::Button(ivec2 pos, ivec2 size, std::string title, Gum::GUI::Font* font)
{
    this->vPos = pos;
    this->vSize = size;
    this->sType = "button";

    backgroundBox = new TextBox(title, font, vec2(0), ivec2(100, 100));
    backgroundBox->setSizeInPercent(true, true);
    addGUI(backgroundBox);

    resize();
    reposition();

    backgroundBox->setTextSize(getSize().y - 10);
}

Button::~Button() 
{
    Gum::_delete(backgroundBox);
}

void Button::update()
{
    if(backgroundBox->isMouseInside())
    {
        backgroundBox->setColor(v4Color - vec4(0.1));
        if(isHoldingLeftClick())
        {
            backgroundBox->setColor(v4Color - vec4(0.3));
        }
        if(!Gum::Window::CurrentlyBoundWindow->getMouse()->isBusy() && isClicked())
        {
            if(callbackFunction != nullptr)
            {
                callbackFunction();
            }
        }
    }
    else
    {
        backgroundBox->setColor(v4Color);
    }
}

void Button::setCallbackFunction(std::function<void()> func)    { this->callbackFunction = func; }
//void Button::setTexture(Texture *newtex)                        { this->backgroundBox->setTexture(newtex); }
void Button::setSecondColor(vec4 col)                           { this->backgroundBox->getBox()->setSecondColor(col); }
void Button::setHasGradient(bool val)                           { this->backgroundBox->getBox()->setHasGradient(val); }
void Button::setGradientDirectionRight(bool val)                { this->backgroundBox->getBox()->setGradientDirectionRight(val); }

//Texture* Button::getTexture()                                   { return this->backgroundBox->getBox()->getTexture(); }
vec4 Button::getSecondColor()                                   { return this->backgroundBox->getBox()->getSecondColor(); }
TextBox *Button::getBox()                                       { return this->backgroundBox; }

Button* Button::createFromXMLNode(XMLNode* node)
{
    Gum::GUI::Font *font = Gum::GUI::Fonts->getDefaultFont();
    float borderRadius = node->mAttributes["radius"]   != "" ? Tools::StringToFloat(node->mAttributes["radius"])   : 0.0f;
    float fontsize     = node->mAttributes["fontsize"] != "" ? Tools::StringToFloat(node->mAttributes["fontsize"]) : 12.0f;
    std::string onclick = "";
    Button *retbutton = new Button(ivec2(0,0), ivec2(1,1), "", font);
    if(onclick != "")
    {
        /*script *currentScript = scripts[onclick];
        if(currentScript == nullptr)
        {
            Gum::Output::error("XMLGUILoader: script '" + onclick + "' does not exist!");
            return retbutton;
        }
        std::function<void()> scriptfunc = std::bind(&script::run, currentScript);
        retbutton->setCallbackFunction(scriptfunc);*/
    }

    return retbutton;
}

File: core/src/main/java/com/gumse/gui/Basics/Button.java
package com.gumse.gui.Basics;

import com.gumse.gui.GUI;
import com.gumse.gui.Locale;
import com.gumse.gui.Font.Font;
import com.gumse.gui.Font.FontManager;
import com.gumse.gui.Primitives.RenderGUI;
import com.gumse.gui.XML.XMLGUI.XMLGUICreator;
import com.gumse.maths.*;
import com.gumse.system.filesystem.XML.XMLNode;
import com.gumse.system.io.Mouse;
import com.gumse.textures.Texture;

public class Button extends RenderGUI
{
    private TextBox backgroundBox;

    public Button(ivec2 pos, ivec2 size, String title, Font font)
    {
        this.vPos.set(pos);
        this.vSize.set(size);
        this.sType = "button";

        backgroundBox = new TextBox(title, font, new ivec2(0), new ivec2(100, 100));
        backgroundBox.setSizeInPercent(true, true);
        backgroundBox.setAlignment(TextBox.Alignment.CENTER);
        backgroundBox.getBox().setBorderThickness(GUI.getTheme().borderThickness);
        backgroundBox.getBox().setCornerRadius(GUI.getTheme().cornerRadius);
        backgroundBox.setTextSize(size.y - 5);
        addGUI(backgroundBox);

        onLeave(new GUICallback() {
            @Override public void run(RenderGUI gui) 
            {
                backgroundBox.setColor(getColor(GUI.getTheme().primaryColor));
            }
        });

        onHover(new GUICallback() {
            @Override public void run(RenderGUI gui) 
            {
                backgroundBox.setColor(vec4.sub(getColor(GUI.getTheme().primaryColor), 0.02f));
                if(isHoldingLeftClick())
                {
                    backgroundBox.setColor(vec4.sub(getColor(GUI.getTheme().primaryColor), 0.05f));
                }
            }
        }, Mouse.GUM_CURSOR_HAND);

        resize();
        reposition();
    }

    @Override
    protected void updateOnTitleChange() 
    {
        backgroundBox.setString(sTitle);
    }

    @Override
    protected void updateOnColorChange() 
    {
        backgroundBox.setColor(getColor(GUI.getTheme().primaryColor));    
    }

    @Override
    protected void updateOnThemeChange() 
    {
        backgroundBox.setColor(getColor(GUI.getTheme().primaryColor));
        backgroundBox.getBox().setBorderThickness(GUI.getTheme().borderThickness);
        backgroundBox.getBox().setCornerRadius(getCornerRadius());
        if(!sLocaleID.isEmpty())
            setTitle(Locale.getCurrentLocale().getString(sLocaleID));
    }

    public void setTexture(Texture newtex)               { this.backgroundBox.setTexture(newtex); }
    public void setSecondColor(vec4 col)                 { this.backgroundBox.getBox().setSecondColor(col); }
    public void setHasGradient(boolean val)              { this.backgroundBox.getBox().setHasGradient(val); }
    public void setGradientDirectionRight(boolean val)   { this.backgroundBox.getBox().setGradientDirectionRight(val); }
    public void setCornerRadius(vec4 radius)             { this.v4CornerRadius = radius; this.backgroundBox.getBox().setCornerRadius(radius); }

    public Texture getTexture()                          { return this.backgroundBox.getBox().getTexture(); }
    public vec4 getSecondColor()                         { return this.backgroundBox.getBox().getSecondColor(); }
    public TextBox getBox()                              { return this.backgroundBox; }

    
    public static XMLGUICreator createFromXMLNode()
    {
        return (XMLNode node) -> { 
            String fontName = node.getAttribute("font");
            Font font = (!fontName.equals("") ? FontManager.getInstance().getFont(fontName) : FontManager.getInstance().getDefaultFont());

            int fontsize     = node.getIntAttribute("fontsize", 0);
            Button retbutton = new Button(new ivec2(0,0), new ivec2(1,1), "", font);
            if(fontsize > 0)
                retbutton.getBox().setTextSize(fontsize);
            

            return retbutton;
        };
    }
};